atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1619,15 +1619,15 @@ protected boolean shouldDeleteAndRollback() {
 
     @Override
     public final void delete(TableReference tableRef, Set<Cell> cells) {
-        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY));
         getCache().delete(tableRef, cells);
+        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY));
```

Comment:
Without this change, there'd still be a race condition:

Thread 1:
-> put internal
-> pause (say GC)

Thread 2 (after put internal from above):
-> read from cache

this would of course fail validation. However, swapping them is safe, as it _just_ invalidates; at the absolute worst this means that we'd read from the remote, so preserves correctness (but in practice this is likely extremely rare).

Reply:
If we swap, and a thread barges in between, we end up seeing in the cache that there has been a local write, but when we try to read it we instead read remote (since it's not cached in the transaction yet). This is ok because we can just pretend that the read happened before the write in this case?
The problem with the old impl is that we would potentially update the cache claiming we read this write remotely because we get confused?

Reply:
I believe the problem with the old impl is that we could update the KVS before the cache, and the read in the middle would fail validation. By swapping the order _and_ changing the write semantics, we guarantee that we either are forced to read remotely or read the locally written value, so it's correct in both cases.